### PR TITLE
Use BlockCompressedInputStream and BlockCompressedOutputStream from htsjdk

### PR DESCRIPTION
**Pull request by robinst**
* Less custom code
* Faster: Indexing the 129 GB file took 1h instead of 1h30m
* More correct: Before, the files we wrote didn't have the right BGZF headers (FLG was 0 instead of 4) and so were not readable by other tools.
* We can take advantage of future improvements such as https://github.com/samtools/htsjdk/pull/1249

Also, `example1b` had an end-of-file-marker in the middle of the file (probably because the blocks were concatenated):

Screen Shot 2020-09-25 at 10 07 00

From spec:

Screen Shot 2020-09-25 at 10 40 13

### CODE REVIEW

File: bam-index/src/main/java/org/victorchang/KeyPointerWriter.java [hunk 1]
```diff
@@ -31,25 +33,23 @@ public KeyPointerWriter() {
 
     @SuppressWarnings("UnstableApiUsage")
     public List<KeyPointer> write(OutputStream outputStream, Stream<KeyPointer> stream, long blockSize) throws IOException {
-        GzipConcatenatedOutputStream concatenatedStream =
-                new GzipConcatenatedOutputStream(outputStream, 1 << 16, compressionLevel);
+        BlockCompressedOutputStream blockCompressedOutputStream = new BlockCompressedOutputStream(outputStream, (File) null, compressionLevel);
```
Comment (huy) [inline bam-index/src/main/java/org/victorchang/KeyPointerWriter.java:36]: I haven't look at the code of htsjdk `BlockCompressedOutputStream`. I feel too risky to use it right away without running some tests first.

File: bam-index/src/main/java/org/victorchang/KeyPointerReader.java [hunk 2]
```diff
@@ -16,27 +14,14 @@
...
-    public Stream<KeyPointer> read(InputStream inputStream, int offset, GzipEntryEventHandler eventHandler) {
-        InputStream concatenatedStream;
-        try {
-            concatenatedStream = new BufferedInputStream(
-                    new GzipConcatenatedInputStream(inputStream, eventHandler), FILE_BUFF_SIZE);
-        } catch (IOException ioException) {
-            throw new RuntimeException(ioException);
-        }
+    public Stream<KeyPointer> read(InputStream inputStream, int offset) {
+        InputStream concatenatedStream = new BlockCompressedInputStream(inputStream);
```
Comment (amitdev) [inline bam-index/src/main/java/org/victorchang/KeyPointerReader.java]: Nit: variable name

**Review by huy — CHANGES_REQUESTED** [review]
oh i miss it with other

File: bam-reader/src/main/java/org/victorchang/DefaultBamFileReader.java [hunk 3]
```diff
@@ -40,51 +37,40 @@ public long read(Path bamFile, BamRecordHandler handler, long bytesLimit) throws
...
                 }
 
-                GzipEntryPosition position = positionFinder.find(uncompressedStream.getBytesRead() - 4);
-
-                if (position == null) {
-                    throw new IllegalStateException("Can't find start of a gzip entry");
-                }
-
-                if (position.getCompressed() >= bytesLimit) {
+                if (coffset > bytesLimit) {
```
Comment (amitdev) [inline bam-reader/src/main/java/org/victorchang/DefaultBamFileReader.java]: This should be `>=`?

File: bam-reader/src/main/java/org/victorchang/PointerPacker.java [hunk 4]
```diff
@@ -1,5 +1,7 @@
 package org.victorchang;
```
Comment (amitdev) [inline bam-reader/src/main/java/org/victorchang/PointerPacker.java]: This class can be deleted now

File: bam-reader/src/main/java/org/victorchang/DefaultBamFileReader.java (hunk 3 above)
Comment (robinst) [reply bam-reader/src/main/java/org/victorchang/DefaultBamFileReader.java]: 👍 done

**Review by amitdev — APPROVED** [review]

**Comment by amitdev** [comment]
Looks good. Run the exhaustive test to make sure nothing is broken, since it is ignored currently.

**Review by huy — APPROVED** [review]

**Comment by robinst** [comment]
> Looks good. Run the exhaustive test to make sure nothing is broken, since it is ignored currently.

Done!

**Pull request merged**